wave_lang/kernel/wave/utils/graph_utils.py

```diff
@@ -212,6 +212,46 @@ def propagate_placeholders(n: fx.Node) -> fx.Node:
     return n
 
 
+def propagate_loop_carried_vars(n: fx.Node, depth: int = 0) -> fx.Node:
+    """
+    Propagates node through placeholders and loop-carried vars.
+
+    `depth` is the number of iterations to propagate through.
+    For example, if `depth` is 1, then the function will propagate through the
+    loop-carried vars to the previous iteration.
+    If `depth` is 0, then the function will propagate to the loop init_args.
+    """
+    c = get_custom(n)
+    if isinstance(c, IterArg):
+        idx = c.iter_idx
+        if idx is None:
+            return n
+        iterate = c.parent_op()
+        assert isinstance(iterate, Iterate), f"Expected Iterate, but got {iterate}"
```

Comment:
Can you use get_inputs in graph_utils.py to get the init_arg corresponding to the iter_arg?

Reply:
Don't think it will help much as we need to branch based in `depth`